bundles/org.openhab.binding.icalendar/src/main/java/org/openhab/binding/icalendar/internal/logic/CommandTag.java

```diff
@@ -0,0 +1,194 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.icalendar.internal.logic;
+
+import java.util.Arrays;
+import java.util.List;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.library.types.HSBType;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.OpenClosedType;
+import org.eclipse.smarthome.core.library.types.PercentType;
+import org.eclipse.smarthome.core.library.types.PlayPauseType;
+import org.eclipse.smarthome.core.library.types.QuantityType;
+import org.eclipse.smarthome.core.library.types.RewindFastforwardType;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.library.types.UpDownType;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.TypeParser;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is a class that implements a Command Tag that may be embedded in an
+ * Event Description. Valid Tags must follow one of the following forms..
+ *
+ * BEGIN:<itemName>:<targetState>
+ * BEGIN:<itemName>:<targetState>:<authorizationCode>
+ * END:<itemName>:<targetState> END:<itemName>:<targetState>:<authorizationCode>
+ *
+ * @author Andrew Fiddian-Green - Initial contribution
+ *
+ */
+@NonNullByDefault
+public class CommandTag {
+    private static final List<Class<? extends Command>> otherCommandTypes = Arrays.asList(QuantityType.class,
+            OnOffType.class, OpenClosedType.class, UpDownType.class, HSBType.class, PlayPauseType.class,
+            RewindFastforwardType.class, StringType.class);
+    private static final List<Class<? extends Command>> percentCommandType = Arrays.asList(PercentType.class);
+
+    private final @Nullable String authorizationCode;
+    private final String fullTag;
+    private final boolean isValid;
+    private final @Nullable String itemName;
+    private final Logger logger = LoggerFactory.getLogger(CommandTag.class);
+    private final @Nullable CommandTagType tagType;
+    private final @Nullable String targetState;
+
+    public CommandTag(String line) {
+        fullTag = line;
+        boolean seemsValid = false;
+        @Nullable
+        String parsedAuthorizationCode = null;
+        @Nullable
+        String parsedItemName = null;
+        @Nullable
+        CommandTagType parsedTagType = null;
+        @Nullable
+        String parsedTargetState = null;
+        boolean parse = true;
```

Comment:
I disagree with this particular coding pattern, instead you should be throwing an exception and letting the caller of this method deal with reporting and dealing with it.

Reply:
@cpmeister hmm, that would require error handling in many places. Maybe these checks are redundant.

@andrewfg Could you please add some unit tests for broken tags (invalid type and item)? It seems like they are missing. Also the logging for the user may never happen if the broken tags are not added to the list (`Event.java` line 46-49 in combination with `CommandTag.java` line 126).

Reply:
Hi Michael, I have been following this PR so far mostly in “lurker” mode; but I will a) add some tests for broken tags, and b) modify the code for handling (and logging) of broken command tags.

PS as the command tags are written by users, I think we must expect to see all possible flavours of “broken” (including even free text that may or may not be a tag at all); so we must 1) log them (IMHO I also think WARN is the proper level) and, 2) catch them to prevent bad things being sent to the core..

Reply:
Done